### PR TITLE
Virtual io

### PR DESCRIPTION
**Pull request by dstreett**
Switching from templates to virtual functions

### CODE REVIEW

File: Super-Deduper/src/super-deduper.cpp [hunk 1]
```diff
@@ -238,55 +189,84 @@ int main(int argc, char** argv)
...
                 
                 if (gzip_out) {
-                    bi::stream<bi::file_descriptor_sink> out1{fileno(popen(("gzip > " + default_outfiles[0] + ".gz").c_str(), "w")), bi::close_handle};
-                    bi::stream<bi::file_descriptor_sink> out2{fileno(popen(("gzip > " + default_outfiles[1] + ".gz").c_str(), "w")), bi::close_handle};
-                    bi::stream<bi::file_descriptor_sink> out3{fileno(popen(("gzip > " + default_outfiles[2] + ".gz").c_str(), "w")), bi::close_handle};
-                    output_read_map_fastq(read_map, out1, out2, out3);
+                    out_1 = new bi::stream<bi::file_descriptor_sink> {fileno(popen(("gzip > " + default_outfiles[0] + ".gz").c_str(), "w")), bi::close_handle};
+                    out_2 = new bi::stream<bi::file_descriptor_sink> {fileno(popen(("gzip > " + default_outfiles[1] + ".gz").c_str(), "w")), bi::close_handle};
+                    out_3 = new bi::stream<bi::file_descriptor_sink> {fileno(popen(("gzip > " + default_outfiles[2] + ".gz").c_str(), "w")), bi::close_handle};
+                    pe = new PairedEndReadOutFastq(*out_1, *out_2);
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: no naked new, lets use unique_ptr

File: Super-Deduper/src/super-deduper.cpp [hunk 2]
```diff
@@ -238,55 +189,84 @@ int main(int argc, char** argv)
...
+                    if(ser) {
+                        se->write(*ser);
+                    }
+                    else {
+                        throw std::runtime_error("Unkown read found");
+                    }
+                }
+            } 
+
+            delete pe;
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: no deletes, use unique_ptr

File: common/src/ioHandler.h [hunk 3]
```diff
@@ -60,54 +60,91 @@ class InterReadImpl : public InputFastq {
 
 // ### output ###
 
-template <class T, class Impl>
+/*template <class T, class Impl>
```
Comment (joe-angell) [inline common/src/ioHandler.h]: remove comments

File: common/test/testIO.cpp [hunk 4]
```diff
@@ -60,49 +60,55 @@ TEST_F(ReadsTest, parsePairedReadFastq) {
...
     
 TEST_F(ReadsTest, testWriteFastqSingle) {
+    
     std::istringstream in1(readData);
     InputReader<SingleEndRead, SingleEndReadFastqImpl> ifs(in1);
 
     std::ostringstream out1;
-    {
-        OutputWriter<SingleEndRead, SingleEndReadOutFastq> ofs(out1);
+    OutputWriter *se = new SingleEndReadOutFastq(out1);
```
Comment (joe-angell) [inline common/test/testIO.cpp]: just say no to new/delete , :)

**Comment by joe-angell** [comment]
yeah looks much better than the template version, 👍 
just a couple minor comments.  Highly recommend getting in the habit of not using new/delete.

File: common/test/testIO.cpp (hunk 4 above)
Comment (dstreett) [inline common/test/testIO.cpp]: `std::unique_ptr<OutputWriter> se(new SingleEndReadOutFastq(out1));`

This sort of new okay? Or is that a no no as well?

**Comment by dstreett** [comment]
Hey, @joe-angell , want to make sure that I used unique_ptrs right before I merge into develop and master? Thanks for looking over this! :)

**Comment by dstreett** [comment]
Introduced a seg fault somewhere in there - tracking it down now.

File: common/test/testIO.cpp (hunk 4 above)
Comment (joe-angell) [inline common/test/testIO.cpp]: yeah we won't get make_unique until c++17 so thats it for now.

**Comment by joe-angell** [comment]
one source of potential issue is that we pass the ostreams by reference, we might want to use shared_ptrs instead.  If they are destroyed before writing is finished, problems will be had.

**Comment by dstreett** [comment]
Ahhh, yeah, I think that is it. Thanks!

**Comment by dstreett** [comment]
Mkay, @joe-angell , I think I fixed the seg fault. You were spot on with the shared versus unique pointers! I had to change some test up to get everything working all right.

Is it looking okay?

File: Super-Deduper/src/super-deduper.cpp [hunk 5]
```diff
@@ -36,7 +36,7 @@ class dbhash {
 };
 
 typedef std::unordered_map<std::string, size_t> Counter;
-typedef std::unordered_map <boost::dynamic_bitset<>, std::unique_ptr<ReadBase>, dbhash> BitMap;
+typedef std::unordered_map <boost::dynamic_bitset<>, std::shared_ptr<ReadBase>, dbhash> BitMap;
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: do we need shared_ptr here?  I think the map should own the reads.

File: Super-Deduper/src/super-deduper.cpp [hunk 6]
```diff
@@ -238,63 +188,82 @@ int main(int argc, char** argv)
                     load_map(ifp, counters, read_map, start, length);
                 }
             }
+            std::unique_ptr<OutputWriter> pe = nullptr;
+            std::unique_ptr<OutputWriter> se = nullptr;
+            std::shared_ptr<std::ostream> out_1 ;
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: missing assignment here

File: Super-Deduper/src/super-deduper.cpp [hunk 7]
```diff
@@ -238,63 +188,82 @@ int main(int argc, char** argv)
...
                 } else {
-                    // note: mapped file is faster but uses more memory
-                    std::ofstream out1(default_outfiles[0], std::ofstream::out);
-                    std::ofstream out2(default_outfiles[1], std::ofstream::out);
-                    std::ofstream out3(default_outfiles[2], std::ofstream::out);
-                    //bi::stream<bi::mapped_file_sink> out1{default_outfiles[0].c_str()};
-                    //bi::stream<bi::mapped_file_sink> out2{default_outfiles[1].c_str()};
-                    //bi::stream<bi::mapped_file_sink> out3{default_outfiles[2].c_str()};
-                    output_read_map_fastq(read_map, out1, out2, out3);
+                    out_1.reset(new std::ofstream (default_outfiles[0], std::ofstream::out));
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: http://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared

File: common/src/ioHandler.h [hunk 8]
```diff
@@ -6,12 +6,12 @@
...
 #include <boost/iostreams/concepts.hpp>
 #include <boost/algorithm/string.hpp>
-
+#include <iostream>
 // ### input ###
 template <class T, class Impl>
 class InputReader : Impl {
 public:
-    typedef std::unique_ptr<T> value_type;
+    typedef std::shared_ptr<T> value_type;
```
Comment (joe-angell) [inline common/src/ioHandler.h]: here again i think we want to keep as unique_ptr.  Unique_ptr has less overhead than shared, and if you aren't sharing the pointer there isn't need for shared.  So think of it like unique_ptr owns an object, and you can use std::move to transfer ownership.  Shared pointer for when multiple things might need to own the object at the same time.

File: common/test/testIO.cpp [hunk 9]
```diff
@@ -58,52 +60,94 @@ TEST_F(ReadsTest, parsePairedReadFastq) {
...
     std::istringstream in1(readData);
     InputReader<SingleEndRead, SingleEndReadFastqImpl> ifs(in1);
 
-    std::ostringstream out1;
-    {
-        OutputWriter<SingleEndRead, SingleEndReadOutFastq> ofs(out1);
+    std::shared_ptr<std::ostream> out1;
+    std::stringbuf test_str;
+    out1.reset(new std::ostringstream);
+    out1->rdbuf(&test_str);
```
Comment (joe-angell) [inline common/test/testIO.cpp]: why are you setting a new rdbuf?  I don't think this is necessary

File: common/test/testIO.cpp [hunk 10]
```diff
@@ -58,52 +60,94 @@ TEST_F(ReadsTest, parsePairedReadFastq) {
...
+
 TEST_F(ReadsTest, testWriteFastqSingle) {
+    
     std::istringstream in1(readData);
     InputReader<SingleEndRead, SingleEndReadFastqImpl> ifs(in1);
 
-    std::ostringstream out1;
-    {
-        OutputWriter<SingleEndRead, SingleEndReadOutFastq> ofs(out1);
+    std::shared_ptr<std::ostream> out1;
```
Comment (joe-angell) [inline common/test/testIO.cpp]: std::shared_ptr<st::ostream> out1 = std::make_shared<std::ostringstream>();  something like that

File: common/test/testIO.cpp [hunk 11]
```diff
@@ -58,52 +60,94 @@ TEST_F(ReadsTest, parsePairedReadFastq) {
...
+    std::stringbuf test_str;
+    out1.reset(new std::ostringstream);
+    out1->rdbuf(&test_str);
+
+    std::unique_ptr<OutputWriter> se(new SingleEndReadOutFastq(out1));
```
Comment (joe-angell) [inline common/test/testIO.cpp]: so I put the outputwriter in a subscope before so it is destroyed (and everything is finished writing) before checking that it is wrtten to the ostringstream.  I suspect that is your issue here, you should put the scope back in place.

File: common/test/testIO.cpp [hunk 12]
```diff
@@ -58,52 +60,94 @@ TEST_F(ReadsTest, parsePairedReadFastq) {
...
+    std::stringbuf test_str_1, test_str_2;
+    out1.reset(new std::ostringstream);
+    out2.reset(new std::ostringstream);
+    out1->rdbuf(&test_str_1);
+    out2->rdbuf(&test_str_2);
+
+    std::unique_ptr<OutputWriter> pe(new PairedEndReadOutFastq(out1, out2));
+
+    {
+        while(ifs.has_next()) {
```
Comment (joe-angell) [inline common/test/testIO.cpp]: put the outputWriter inside this scope, so flush is called when the outputwriter is destoryed.  Then you don't need all the extra stringbuf stuff etc.

**Comment by joe-angell** [comment]
looking close, need to clean up the tests.  So remember the destructor of outputwriter flushes the ostream, so we can't be sure the data has been written until after that is destroyed.  Clean up the unit tests a little and you should be pretty good.  

Oh and c++14 has make_unique, so we might be able to upgrade to that (although i think sam said the cluster has an older version of g++ so we may need to stay on c++11).

File: Super-Deduper/src/super-deduper.cpp [hunk 13]
```diff
@@ -238,63 +188,82 @@ int main(int argc, char** argv)
                     load_map(ifp, counters, read_map, start, length);
                 }
             }
+            std::unique_ptr<OutputWriter> pe = nullptr;
+            std::unique_ptr<OutputWriter> se = nullptr;
+            std::shared_ptr<std::ostream> out_1 ;
+            std::shared_ptr<std::ostream> out_2 = nullptr;
+            std::shared_ptr<std::ostream> out_3 = nullptr;
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: Note that c++ will destroy things on the stack in reverse order that they are created.  So when you first switched to shared_ptr, I bet you had this order, which will delete the ostreams before outputwriter, and we need the opposite.  Now that they are shared_ptr that will be enforced though.

**Comment by dstreett** [comment]
Sorry, @joe-angell, I am a total newb. -_- As soon as you said reverse stack order, it totally made sense about the seg fault. We can keep pretty much all the code the same, just change that ordering.

Sorry, to make you look over those 100% unnecessary changes and waste your time. I did learn a lot though. :) 

Do you think it is worth it to put the ostream shared_ptrs in OutputWriter? We actually don't need them.

**Comment by joe-angell** [comment]
@dstreett no need to be sorry.  I enjoy teaching what I can to a receptive person, and I'm learning too, ;).  As for changing OutputWriter to take ostream shared_ptrs, I leave it to you, i'm happy either way.   It might be nice to leave just so you get practice with shared_ptr/unique_ptr and their usage.

**Comment by dstreett** [comment]
Thanks, @joe-angell . I do appreciate it! :) I say let's get this all merged in with the new writer, and I will make a separate branch with shared pointers throughout. That shouldn't change the interface to write at all, so development can keep happening on the different applications. Sound good?

**Comment by joe-angell** [comment]
ok sounds good

**Comment by dstreett** [comment]
Do to merge, do you think?

File: common/test/testIO.cpp [hunk 14]
```diff
@@ -60,16 +60,17 @@ TEST_F(ReadsTest, parsePairedReadFastq) {
...
     
 TEST_F(ReadsTest, testWriteFastqSingle) {
+    
     std::istringstream in1(readData);
     InputReader<SingleEndRead, SingleEndReadFastqImpl> ifs(in1);
 
     std::ostringstream out1;
-    {
-        OutputWriter<SingleEndRead, SingleEndReadOutFastq> ofs(out1);
+    std::unique_ptr<OutputWriter> se(new SingleEndReadOutFastq(out1));
```
Comment (joe-angell) [inline common/test/testIO.cpp]: this should be inside the { context on line 70, so the output writer is destroyed (and flush is called) before the assert.

File: common/test/testIO.cpp [hunk 15]
```diff
@@ -78,31 +79,35 @@ TEST_F(ReadsTest, testWriteFastqSingle) {
 TEST_F(ReadsTest, testTabWrite) {
     std::istringstream in1(readTabData);
     InputReader<ReadBase, TabReadImpl> ifs(in1);
-    
     std::ostringstream out1;
+
+    std::unique_ptr<OutputWriter> ofs(new ReadBaseOutTab(out1));
```
Comment (joe-angell) [inline common/test/testIO.cpp]: same here

File: common/test/testIO.cpp [hunk 16]
```diff
@@ -78,31 +79,35 @@ TEST_F(ReadsTest, testWriteFastqSingle) {
...
+
     ASSERT_EQ(readTabData, out1.str());
 }
 
 TEST_F(ReadsTest, testInterWrite) {
     std::istringstream in1(readDataInter);
     InputReader<PairedEndRead, InterReadImpl> ifs(in1);
     std::ostringstream out1;
+
+    std::unique_ptr<OutputWriter> ofs(new PairedEndReadOutInter(out1));
```
Comment (joe-angell) [inline common/test/testIO.cpp]: and here

**Comment by joe-angell** [comment]
yeah fix those minor issues in the unit tests and it should be good to merge

**Pull request merged**